### ISSUE TITLE
Allow configuration of React.Component subclasses in settings

### ISSUE THREAD

**Issue by stevoland** [yannickcr/eslint-plugin-react#717]
Hi, apologies if this has been addressed before. I know subclassing `React.Component` is frowned upon but we make extensive use of a couple of subclasses for historical reasons.

I wonder if you'd consider accepting a PR to allow configuring Regexes of module paths which export these subclasses. Something like:

```json
{
  "settings": {
    "react": {
      "componentSubclassMatchers": [{
        "source": "Shared",
        "name": "BaseComponent"
      }]
    }
  }
}
```

To allow the following as `isES6Component`:

```javascript
import {BaseComponent} from 'SharedModule';

class AComponent extends BaseComponent {}
```

I realise this can be addressed with `/* @extends React.Component */` but then this approach requires another rule to enforce this doc tag in order to enforce the react rules. Thanks

**Comment by wrakky** [comment]
I've just come across this problem myself. It would be good to know some thoughts on the matter - I am also be happy to dedicate some time to this if it hasn't been rejected.

**Comment by lencioni** [comment]
I'd prefer to not add complexity to support such a thing, but since it seems like this would help you out, I think something like this would be acceptable. I think it would be good to mention in the documentation that subclassing `React.Component` should be avoided.

**Comment by ljharb** [comment]
There's no need for additional configuration - using JSDoc-style "@extends React.Component" (or something similar, I forget the exact syntax) can inform eslint-plugin-react that something subclasses a React component.

**Comment by lencioni** [comment]
Ah, interesting. Is this documented anywhere?

**Comment by ljharb** [comment]
No, I don't believe so - precisely because people shouldn't need it.

**Comment by AMongeEntytle** [comment]
Could someone explain with some extra details how can this "extends" rule should be defined? Is it necessary to include any extra dependencies to make eslint detect and 'understand' jsdoc rules? 
Thank you !

**Comment by yannickcr** [comment]
You do not have to add any extra dependencies, just add the JSDoc-style comment before your component and `eslint-plugin-react` should be able to detect your component even if you do not extend `React.Component` directly:

```jsx
/** @extends React.Component */
class Hello extends ChildComponent {
  render() {
    return <div>Hello {this.props.name}</div>;
  }
}
```

This possibility should be added to the documentation.

**Comment by AMongeEntytle** [comment]
@yannickcr I finally got it working by myself (after a little bit of digging into the code :P) but I think is also good that this is, at least, here in this issue. (and yes, you might want to include it in the documentation =D)
Thank you for your work !
